### PR TITLE
Remove Feature Flag Service in favor of OpenFeature + flagd

### PR DESCRIPTION
**Pull request by austinlparker**
# Changes

This PR removes the Erlang featureflag service and its dependencies in favor of [OpenFeature](https://openfeature.dev) ecosystem projects.

## Rationale

Feature Flag Service is a bit of a heavy dependency for the project to maintain. We don't have any regular Erlang contributors in maintainer/approver ranks, which makes it difficult to evaluate changes. There's been a fairly long-standing issue with cross-platform builds in QEMU and various Erlang components which has kept the service on outdated versions of OTP. We've had several feature requests for more flexibility around the feature flags that require setting via API, and the lack of a dedicated client for the feature flag service itself makes it somewhat challenging for contributors to add new things (as they have to write handlers for each application).

OpenFeature is a CNCF incubating project that's under active development and provides several things:
- Client libraries for most languages
- Lightweight flag daemon (flagd)
- Kubernetes integration
- Native OpenTelemetry support

## Merge Requirements

For new features contributions please make sure you have completed the following
essential items:

* [x] `CHANGELOG.md` updated to document new feature additions
* [x] Appropriate documentation updates in the [docs][]
* [ ] Appropriate Helm chart updates in the [helm-charts][]

Maintainers will not merge until the above have been completed. If you're unsure
which docs need to be changed ping the
[@open-telemetry/demo-approvers](https://github.com/orgs/open-telemetry/teams/demo-approvers).

[docs]: https://opentelemetry.io/docs/demo/
[helm-charts]: https://github.com/open-telemetry/opentelemetry-helm-charts

### CODE REVIEW

**Comment by julianocosta89** [comment]
Wow!
This is a huge change but I'd hold it for a bit.
I wouldn't replace the FeatureFlagService without having another Erlang/Elixir service on the demo.

**Comment by austinlparker** [comment]
> Wow! This is a huge change but I'd hold it for a bit. I wouldn't replace the FeatureFlagService without having another Erlang/Elixir service on the demo.

See https://github.com/open-telemetry/opentelemetry-demo/issues/1343#issuecomment-1952757370 -- I think we'll need to get another Erlang service added in as a stub off Kafka somehow. I don't necessarily think it should block this PR, though.

**Comment by julianocosta89** [comment]
> > Wow! This is a huge change but I'd hold it for a bit. I wouldn't replace the FeatureFlagService without having another Erlang/Elixir service on the demo.
> 
> See [#1343 (comment)](https://github.com/open-telemetry/opentelemetry-demo/issues/1343#issuecomment-1952757370) -- I think we'll need to get another Erlang service added in as a stub off Kafka somehow. I don't necessarily think it should block this PR, though.

So you would drop Erlang/Elixir from the demo till we get a new service rewritten?

**Comment by julianocosta89** [comment]
I vote against it.

**Comment by austinlparker** [comment]
We can discuss it at the next SIG meeting. I don't anticipate releasing the demo without an Erlang service, but we also can't arbitrarily hold off on integrating new changes because of implicit dependencies.

**Comment by tsloughter** [comment]
Wish I'd known about this I'd have done what I could to make it not necessary, I wasn't aware it was being considered. Does Erlang/Elixir now get no demo app?

**Comment by austinlparker** [comment]
> Wish I'd known about this I'd have done what I could to make it not necessary, I wasn't aware it was being considered. Does Erlang/Elixir now get no demo app?

No, we'll still have an Erlang demo app, but probably a smaller one that's less of a maintenance burden?

**Comment by tsloughter** [comment]
So I can try to help, what burdern does it ease to be another service? You mention the build pipeline, how would that be different for product service or the new kafka service?

**Comment by austinlparker** [comment]
@tsloughter moving this convo over to #1343

**Comment by beeme1mr** [comment]
Hey @austinlparker, I'm one of the maintainers of OpenFeature. Thanks for getting this started! Please let me know if you have any questions.

**Comment by agardnerIT** [comment]
> I'd have done what I could to make it not necessary

For the future, calls to the OpenFeature SDK can specify a default value so even if the backend is unavailable, you'll still get a working app.

**Comment by austinlparker** [comment]
> > I'd have done what I could to make it not necessary
> 
> 
> 
> For the future, calls to the OpenFeature SDK can specify a default value so even if the backend is unavailable, you'll still get a working app.

Feel free to check my work but I'm pretty sure that's what I've done here.

**Comment by austinlparker** [comment]
> Hey @austinlparker, I'm one of the maintainers of OpenFeature. Thanks for getting this started! Please let me know if you have any questions.

Rad! Could you look into getting the python flagd provider released? :)

**Comment by beeme1mr** [comment]
> Rad! Could you look into getting the python flagd provider released? :)

@thisthat has opened a PR. Hopefully, we can get publishing working tomorrow.

https://github.com/open-feature/python-sdk-contrib/pull/38

**Comment by austinlparker** [comment]
Perfect!

**Comment by beeme1mr** [comment]
We've published the Python flagd provider. You can see it [here](https://pypi.org/project/openfeature-provider-flagd/0.1.2/).

**Comment by austinlparker** [comment]
Seems like there's something up with it.. I filed an issue. https://github.com/open-feature/python-sdk-contrib/issues/44

**Comment by beeme1mr** [comment]
> Seems like there's something up with it.. I filed an issue. [open-feature/python-sdk-contrib#44](https://github.com/open-feature/python-sdk-contrib/issues/44)

Sorry about that! I posted here as soon as the publishing was successful, but I should have verified that it was working first. We'll look into the issue as soon as possible.

**Comment by austinlparker** [comment]
New error on 0.1.3 -- https://github.com/open-feature/python-sdk-contrib/issues/48

**Comment by austinlparker** [comment]
User error on that 0.1.3 issue, I also needed to update the SDK. Will update this PR momentarily.

**Comment by agardnerIT** [comment]
The `checkAdFailure` logic could, if you want, move into flagd?

This will return `false` if and only if the number you pass is `1`

```
{
  "flags": {
    "adServiceFailure": {
      "state": "ENABLED",
      "defaultVariant": "false",
      "variants": {
        "true": true,
        "false": false
      },
      "targeting": {
        "if": [
          {
            "==": [{ "var": "number" }, 1 ]
          },
          "true"
        ]
      }
    }
  }
}
```

For example:

```
eval_context = EvaluationContext(
    targeting_key="targeting_key2", attributes={"number": 123}
client.getBooleanValue("adServiceFailure", false, eval_context)
)
```

File: src/adservice/build.gradle
Comment (justinabrahms) [inline src/adservice/build.gradle:54]:
```suggestion
                "dev.openfeature.contrib.providers:flagd:0.7.0",
                'dev.openfeature:sdk:1.7.4'
```

File: src/adservice/src/main/java/oteldemo/AdService.java
Comment (justinabrahms) [inline src/adservice/src/main/java/oteldemo/AdService.java:198]: nitpick, but I'd move this up so you don't fetch a flag value if you don't need to.

**Review by justinabrahms — APPROVED** [review]
Approving as a maintainer of the java openfeature stuff. Thanks! :)

File: src/adservice/src/main/java/oteldemo/AdService.java
Comment (austinlparker) [reply src/adservice/src/main/java/oteldemo/AdService.java]: I changed this up so that we do the fractional eval in flagd rather than in the code.

**Comment by austinlparker** [comment]
@open-telemetry/demo-maintainers could I get a review here?

**Comment by austinlparker** [comment]
I can handle the docs updates if someone else can do Helm.

**Comment by basti1302** [comment]
I only just found out about this, otherwise I would have joined the discussion sooner. The initial PR description mentions outstanding feature requests like "setting [feature flags] via API", and I wanted to at least mention that I have worked on that recently and implemented it in [Dash0's fork](https://github.com/dash0hq/opentelemetry-demo). On that fork, the usual CRUD operations for feature flags are available via the frontend API, so stuff like `curl  -X PUT -H "Content-Type: application/json" https://$your-otel-demo-base-url/api/featureflags/adServiceFailure --data "{\"value\": 0.5}"` works.

In addition to the error rates the feature flag service currently mostly supports, I also added support for simulated slowness of requests, which are based on three settings persisted as feature flag (probability of slow requests, minimum and maximum delay).

I had planned to contribute those changes upstream in the next few days. Obviously, these changes would become obsolete if this PR is merged (or vice versa).

I wanted to at least bring it up as an alternative to this PR. The other factor that has been mentioned a couple of times is the fact that when this PR is merged, there is no Erlang/Elixir service left in the demo, which might speak in favor of keeping the current feature flag service implementation and extending it. I can definitely help with the maintenance of the Erlang code.

Of course, substantial work has probably already gone into this PR and there are certainly other aspects in favor of using a more fully fledged solution like OpenFeature/flagd instead of a DIY thing like the current feature flag service.

I don't have a strong opinion either way, but as I said, I wanted to at list bring this option to the table.

**Comment by austinlparker** [comment]
@basti1302 ah, yeah, that sucks. Unfortunately, part of the goal here is to remove the erlang/elixir service from the critical path of the app entirely due to various issues around maintenance and builds and so forth, not to mention supporting other CNCF ecosystem projects like OpenFeature. I also feel like we don't want to pack too much extraneous logic (such as the simulated slowness/delays) into custom components; I would rather us use chaos frameworks to inject latency.

I do appreciate the work you've done in those PRs, though. Perhaps you could help out on creating a replacement erlang service as discussed in https://github.com/open-telemetry/opentelemetry-demo/issues/1343

**Comment by tsloughter** [comment]
@basti1302 nice, thanks, wish it would have worked out. I would fight this if it wasn't for the QEMU issue on M1+ -- and now I think at this point it is too far along to be worth finding out if the issue is resolved in OTP-27-rc.

**Comment by mviitane** [comment]
When running this branch, I constantly get this error:
`frontend                 | Error: 12 UNIMPLEMENTED: Method not found: oteldemo.AdService/GetAds`
Anyone else see this?

**Comment by julianocosta89** [comment]
@mviitane yes!

```
Error: 12 UNIMPLEMENTED: Method not found: oteldemo.AdService/GetAds
    at callErrorFromStatus (/app/node_modules/@grpc/grpc-js/build/src/call.js:31:19)
    at Object.onReceiveStatus (/app/node_modules/@grpc/grpc-js/build/src/client.js:192:76)
    at Object.onReceiveStatus (/app/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:360:141)
    at Object.onReceiveStatus (/app/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:323:181)
    at /app/node_modules/@grpc/grpc-js/build/src/resolving-call.js:99:78
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
for call at
    at ServiceClientImpl.makeUnaryRequest (/app/node_modules/@grpc/grpc-js/build/src/client.js:160:32)
    at ServiceClientImpl.<anonymous> (/app/node_modules/@grpc/grpc-js/build/src/make-client.js:105:19)
    at /app/node_modules/@opentelemetry/instrumentation-grpc/build/src/grpc-js/clientUtils.js:131:31
    at /app/node_modules/@opentelemetry/instrumentation-grpc/build/src/grpc-js/index.js:233:209
    at AsyncLocalStorage.run (node:async_hooks:338:14)
    at AsyncLocalStorageContextManager.with (/app/node_modules/@opentelemetry/context-async-hooks/build/src/AsyncLocalStorageContextManager.js:33:40)
    at ContextAPI.with (/app/node_modules/@opentelemetry/api/build/src/api/context.js:60:46)
    at ServiceClientImpl.clientMethodTrace [as getAds] (/app/node_modules/@opentelemetry/instrumentation-grpc/build/src/grpc-js/index.js:233:42)
    at /app/.next/server/pages/api/data.js:63:58
    at new Promise (<anonymous>) {
  code: 12,
  details: 'Method not found: oteldemo.AdService/GetAds',
  metadata: Metadata {
    internalRepr: Map(1) { 'content-type' => [Array] },
    options: {}
  }
}
```

@austinlparker was that working on the initial PR?

**Comment by julianocosta89** [comment]
Ads are not being loaded at all:
image

**Comment by austinlparker** [comment]
Huh. I thought it was?

**Comment by austinlparker** [comment]
Hm, does seem like it's not working. The ad service is running fine though.

**Comment by austinlparker** [comment]
Should be working now, my bad.

**Comment by austinlparker** [comment]
@beeme1mr looks like the python otel hook isn't published yet?

File: src/flagd/demo.flagd.json
Comment (austinlparker) [reply src/flagd/demo.flagd.json]: 👍🏼

**Review by julianocosta89 — APPROVED** [review]
Tested it, and everything seems to be working fine.
Let's ship it! 🚀

**Comment by beeme1mr** [comment]
> @beeme1mr looks like the python otel hook isn't published yet?

We'll publish the hook and include it in a follow up PR.

**Comment by federicobond** [comment]
@austinlparker the Python OTel hook is now released! Note that you need to update the import paths for the released version

**Comment by julianocosta89** [comment]
@federicobond I've tried adding it, but it seems there is an issue with the hook.
I've opened an issue here: https://github.com/open-feature/python-sdk-contrib/issues/67

**Pull request merged**
